### PR TITLE
aws-greengrass-core-sdk basic definitions.

### PR DESCRIPTION
**Pull request by ajhattozxcorp**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## aws-greengrass-core-sdk ([<kbd>unpkg</kbd>](https://unpkg.com/browse/aws-greengrass-core-sdk@latest/))
was missing the following properties:
1. GreengrassInterfaceVersion
1. Lambda
1. SecretsManager
1. StreamManager
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against c26311404b4a1f361d32cb58e064b49fc620b0f5
</p>

File: types/aws-greengrass-core-sdk/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    queueFullPolicy?: 'AllOrError' | 'BestEffort';
+}
+
+export interface IotCallbackData {
+    payload: Buffer | Blob | string | ArrayLike<any>;
+}
+
+export type IotCallback = (error: Error | null, data: IotCallbackData | null) => void;
+
+export interface IotDataClient {
```
Comment (amcasey) [inline types/aws-greengrass-core-sdk/index.d.ts:20]: I'm not sure I understand what this adds.  Couldn't it be replaced by `IotData` at any use site?

**Review by amcasey — APPROVED** [review]
Quick question, but also fine to merge.

File: types/aws-greengrass-core-sdk/index.d.ts (hunk 1 above)
Comment (ajhattozxcorp) [reply types/aws-greengrass-core-sdk/index.d.ts:20]: You're correct. I had defined that interface in my use project for unit/integration testing, but it doesn't really belong in the definition library as clients could easily implement their own interface to the concrete class as required. 

I'll make an update.

Comment (ajhattozxcorp) [reply types/aws-greengrass-core-sdk/index.d.ts]: Done. Not sure why CI is failing though. :(

Comment (amcasey) [reply types/aws-greengrass-core-sdk/index.d.ts]: The CI failure is unrelated.  Thanks for the update!

**Pull request merged**